apps/src/lib/kits/maker/boards/microBit/MBFirmataUpdater.js

```diff
@@ -0,0 +1,91 @@
+import {
+  MICROBIT_VENDOR_ID,
+  MICROBIT_PRODUCT_ID,
+  MICROBIT_FIRMATA_V1_URL,
+  MICROBIT_FIRMATA_V2_URL,
+  MICROBIT_IDS_V1,
+  MICROBIT_IDS_V2,
+  MICROBIT_V1,
+  MICROBIT_V2
+} from '@cdo/apps/lib/kits/maker/boards/microBit/MicroBitConstants';
+import {DAPLink, WebUSB} from 'dapjs';
+import {getStore} from '@cdo/apps/redux';
+import {setMicroBitFirmataUpdatePercent} from '@cdo/apps/lib/kits/maker/microBitRedux';
+
+export default class MBFirmataUpdater {
+  constructor() {
+    // State value 'microbitFirmataUpdatePercent' is only updated when this instance variable changes.
+    this.firmataUpdatePercent = 0;
+  }
+
+  detectMicroBitVersion(device) {
+    // Detect micro:bit version V1 or V2 from the first 4 digits of the micro:bit's serial number
+    // Documentation at https://support.microbit.org/support/solutions/articles/19000035697-what-are-the-usb-vid-pid-numbers-for-micro-bit
+    const microBitId = device.serialNumber.substring(0, 4);
+    let microBitVersion = null;
+    if (MICROBIT_IDS_V1.includes(microBitId)) {
+      microBitVersion = MICROBIT_V1;
+    } else if (MICROBIT_IDS_V2.includes(microBitId)) {
+      microBitVersion = MICROBIT_V2;
+    }
+    return microBitVersion;
+  }
+
+  // Precondition: microBitVersion is assigned either MICROBIT_V1 or MICROBIT_V2
```

Comment:
Just to be safe, if you want to enforce this precondition, you can `throw` if microbitVersion isn't V1 or V2 (and unit test that as well).

Reply:
Nice! Just added `throw` and corresponding unit test.